### PR TITLE
Remove automatic buffering in `ipc::reader::FileReader` for for consistent buffering

### LINKED ISSUE #6099: Allow flushing or non-buffered writes from `arrow::ipc::writer::StreamWriter`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I am designing an API boundary that uses Arrow IPC to pass messages around. My requirement is that I need control at granularity of individual encapsulated messages, so for example I want to be able to force that at a given moment the initial Schema has definitely been written to the stream, or that all RecordBatches written thus far have been flushed.

I am using a custom `Writer` implementation with `StreamWriter`. Currently this is impossible to achieve because the `StreamWriter` [internally wraps any passed `Writer` into a `BufWriter`](https://github.com/apache/arrow-rs/blob/ee5694078c86c8201549654246900a4232d531a9/arrow-ipc/src/writer.rs#L1035) and exposes no way to flush.

**Describe the solution you'd like**
The simplest solution would be to expose an `fn flush(&mut self)` on the `StreamWriter` that would force the internal `BufWriter` to flush to the user-provided `Writer`.

**Describe alternatives you've considered**
As an alternative and a bigger (IMO) improvement, an option could be given to `IpcWriteOptions` to disable internal buffering. In my case the custom `Writer` implementation is quite complex and well-optimised already, and the `BufWriter` on top does not actually give any benefit, rather creates an additional overhead. A simple flag `use_buf_writer` set to `true` by default but possible to override would solve this. Internally the `writer` field of `StreamWriter` would need to be changed to some more complicated writer, e.g.

```rust
enum MaybeBuffered<W> {
  Raw(W),
  Buffered(BufWriter<W>)
}
```

EDIT: I looked more into the codebase and I can see there is `try_new_unbuffered` in `StreamReader` already, but that API is a bit different - the "buffered" version explicitly returns `StreamReader<BufReader<R>>`, so this wouldn't be possible to do for the writer API without a breaking change.

I'd be willing to make a PR for either of those changes, but wanted to ask you whether these make sense and would be accepted. This is quite important and blocking for me at the moment.

### PR DESCRIPTION
**Pull request by V0ldek**
# Which issue does this PR close?

Closes #6099.

# Rationale for this change
 

Current writer API automatically wraps the supplied `std::io::Writer` impl into a `BufWriter`.
It is cleaner and more idiomatic to have the default be using the supplied impl directly, as the user might already have a `BufWriter` or an impl that doesn't actually benefit from buffering at all.

StreamReader does a similar thing, but it also exposes a `try_new_unbuffered` that bypasses the internal wrap. These APIs are incosistent between reader/writer and also between `Stream` and `File` versions, but there's no strong reason for that.

# What changes are included in this PR?

Here we propose a consistent and non-buffered by default API:
- `try_new` does not wrap the passed reader/writer,
- `try_new_buffered` is a convenience function that does wrap the reader/writer into a BufReader/BufWriter,
- all four publicly exposed IPC reader/writers follow the above consistently, i.e. `StreamReader`, `FileReader`, `StreamWriter`, `FileWriter`.

An additional tweak: removed the generic type bounds from struct definitions on the four types, as that is the idiomatic Rust approach (see e.g. stdlib's `HashMap` that has no bounds on the struct definition, only the impl requires `Hash + Eq`).

# Are there any user-facing changes?

- [BREAKING] function `arrow_ipc::reader::StreamReader<R>::try_new_unbuffered` is deprecated to be removed in a future version;
- [BREAKING] function `arrow_ipc::reader::StreamReader<std::io::BufReader<R>>::try_new` is renamed to `try_new_buffered`;
- function `arrow_ipc::reader::FileReader<std::io:::BufReader<R>>::try_new_buffered` is added;
- function `arrow_ipc::writer::StreamWriter<std::io::BufWriter<W>>::try_new_buffered` is added;
- function `arrow_ipc::writer::FileWriter<std::io::BufWriter<W>>::try_new_buffered` is added.

The fact that `try_new` is now unbuffered is a silent change from the API perspective. A number of rustdocs explicitly referenced the internal buffering and have been reworded accordingly.

### CODE REVIEW

**Comment by V0ldek** [comment]
This is a breaking change and should be labelled with `api-change`, but I can't add it as an external contributor :)

File: arrow-ipc/src/reader.rs [hunk 1]
```diff
@@ -1163,21 +1172,24 @@ impl<R: Read> fmt::Debug for StreamReader<R> {
...
 impl<R: Read> StreamReader<R> {
-    /// Try to create a new stream reader but do not wrap the reader in a BufReader.
+    /// Try to create a new stream reader.
+    ///
+    /// The first message in the stream is the schema, the reader will fail if it does not
+    /// encounter a schema.
+    /// To check if the reader is done, use [`is_finished(self)`](StreamReader::is_finished).
     ///
-    /// Unless you need the StreamReader to be unbuffered you likely want to use `StreamReader::try_new` instead.
-    pub fn try_new_unbuffered(
```
Comment (alamb) [inline arrow-ipc/src/reader.rs:1180]: To ease migration, what do you think about  adding a `try_new_unbuffered` function (that calls `try_new`) marked as `deprecated` ?

File: arrow-ipc/src/writer.rs [hunk 2]
```diff
@@ -844,6 +844,15 @@ pub struct FileWriter<W: Write> {
...
+    /// Try to create a new file writer with the writer wrapped in a BufWriter.
+    ///
+    /// See [`FileWriter::try_new`] for an unbuffered version.
+    pub fn try_new_buffered(writer: W, schema: &Schema) -> Result<Self, ArrowError> {
+        Self::try_new(BufWriter::new(writer), schema)
+    }
+}
+
 impl<W: Write> FileWriter<W> {
     /// Try to create a new writer, with the schema written as part of the header
```
Comment (alamb) [inline arrow-ipc/src/writer.rs:857]:
```suggestion
    /// Try to create a new writer, with the schema written as part of the header
    ///
    /// Note the created writer is not buffered. See [`FileWriter::try_new_buffered`] for details.
```

File: arrow-ipc/src/writer.rs [hunk 3]
```diff
@@ -853,12 +862,11 @@ impl<W: Write> FileWriter<W> {
 
     /// Try to create a new writer with IpcWriteOptions
```
Comment (alamb) [inline arrow-ipc/src/writer.rs:863]:
```suggestion
    /// Try to create a new writer with IpcWriteOptions
    ///
    /// Note the created writer is not buffered. See [`FileWriter::try_new_buffered`] for details.
```

File: arrow-ipc/src/writer.rs [hunk 4]
```diff
@@ -990,16 +998,20 @@ impl<W: Write> FileWriter<W> {
...
-    /// writer
+    /// Unwraps the the underlying writer.
+    ///
+    /// The writer is flushed and the FileWriter is finished before returning.
     ///
-    /// The buffer is flushed and the FileWriter is finished before returning the
-    /// writer.
+    /// # Errors
+    ///
+    /// An ['Err'](Result::Err) may be returned if an error occurs while finishing the StreamWriter
```
Comment (alamb) [inline arrow-ipc/src/writer.rs:1007]: 👍

**Review by alamb — APPROVED** [review]
Thank you @V0ldek  -- I think this is a nice improvement. I had some small comment suggestions, but we could potentially do them as follow on PRs

File: arrow-ipc/src/writer.rs (hunk 4 above)
Comment (V0ldek) [reply arrow-ipc/src/writer.rs:1007]: I'm going to add those Errors sections to the `try_new*` functions as well in hindisght.

File: arrow-ipc/src/reader.rs (hunk 1 above)
Comment (V0ldek) [reply arrow-ipc/src/reader.rs:1180]: Definitely no harm in that :)

File: arrow-ipc/src/reader.rs [hunk 5]
```diff
@@ -1224,6 +1246,15 @@ impl<R: Read> StreamReader<R> {
         })
     }
 
+    /// Deprecated, use [`StreamReader::try_new`] instead.
```
Comment (alamb) [inline arrow-ipc/src/reader.rs:1249]: 👍

**Review by alamb — APPROVED** [review]
Thank you @V0ldek  -- I think this is a very nice improvement

**Pull request merged**